File: aoc22/py/day02.py
#!/usr/bin/env python

from utils import inputs, examples


def match_score(opp, you):
    # Loss = 0, Win = 6, draw = 3
    diff = (opp - you) % 3
    if diff == 0:
        return 3
    elif diff == 2:
        return 6
    elif diff == 1:
        return 0


def score(play):
    opp, you = play
    return 1 + you + match_score(opp, you)


def strategy(opp, strategy):
    # (0, 1, 2) = (lose, draw, win)
    if strategy == 1:
        return (opp, opp)
    elif strategy == 0:
        return (opp, (opp + 2) % 3)
    elif strategy == 2:
        return (opp, (opp + 1) % 3)


def day02(filename):
    print()
    print(filename)

    with open(filename) as puzzlein:
        plays = []
        for line in puzzlein:
            opp, you = line.split()
            play = ord(opp) - ord("A"), ord(you) - ord("X")
            plays.append(play)

        part1 = 0
        part2 = 0
        for play in plays:
            part1 += score(play)
            part2 += score(strategy(*play))

        print("part1:", part1)
        print("part2:", part2)


day02(inputs("02"))
day02(examples("02"))

File: aoc22/src/day02.rs
#![allow(dead_code)]

use std::convert::TryInto;

#[allow(unused_imports)]
use anyhow;

pub fn day02() {
    println!("day02 part1: {}", part1(&input()));
    println!("day02 part2: {}", part2(&input()));
}

#[derive(Debug, Clone, Copy)]
enum RPS {
    Rock = 0,
    Paper = 1,
    Scissors = 2,
}

impl RPS {
    fn fight(&self, other: &Self) -> i32 {
        let diff = (*self as i8) - (*other as i8);
        match diff.rem_euclid(3) {
            0 => 3,
            1 => 0,
            2 => 6,
            i8::MIN..=-1_i8 | 3_i8..=i8::MAX => todo!(),
        }
    }

    fn strategy(&self, strategy: u8) -> (Self, Self) {
        let other = match strategy as char {
            'X' => (*self as i32 + 2).rem_euclid(3).try_into(), // Other should lose against self
            'Y' => Ok(*self),                                   // Draw the matchup
            'Z' => (*self as i32 + 1).rem_euclid(3).try_into(),
            _ => todo!(), // Should not happen
        };
        (*self, other.unwrap())
    }
}

impl TryFrom<i32> for RPS {
    type Error = i32;

    fn try_from(v: i32) -> Result<RPS, i32> {
        match v {
            0 => Ok(RPS::Rock),
            1 => Ok(RPS::Paper),
            2 => Ok(RPS::Scissors),
            _ => Err(v),
        }
    }
}

impl TryFrom<char> for RPS {
    type Error = i32;

    fn try_from(v: char) -> Result<RPS, i32> {
        match v {
            'A' | 'B' | 'C' => ((v as i32) - ('A' as i32)).try_into(),
            'X' | 'Y' | 'Z' => ((v as i32) - ('X' as i32)).try_into(),
            _ => Err(-1),
        }
    }
}

fn score(matchups: Vec<(RPS, RPS)>) -> i32 {
    matchups
        .iter()
        .map(|(opp, you)| 1 + (*you as i32) + opp.fight(you))
        .sum()
}

fn part1(input: &str) -> i32 {
    let matchups: Vec<(RPS, RPS)> = input
        .lines()
        .map(|line| {
            let bytes = line.as_bytes();
            let (opp, you) = (bytes[0] as char, bytes[2] as char);
            (
                opp.try_into().expect("opponent part1"),
                you.try_into().expect("you part1"),
            )
        })
        .collect();
    score(matchups)
}

fn part2(input: &str) -> i32 {
    let matchups: Vec<(RPS, RPS)> = input
        .lines()
        .map(|line| {
            let bytes = line.as_bytes();
            let opp: RPS = (bytes[0] as char).try_into().unwrap();
            opp.strategy(bytes[2])
        })
        .collect();
    score(matchups)
}

fn input() -> String {
    std::fs::read_to_string("inputs/02").expect("inputs/02 to exist")
}

fn example() -> String {
    std::fs::read_to_string("examples/02").expect("examples/02 to exist")
}

#[test]
fn test_part1() {
    assert_eq!(15, part1(&example()));
    assert_eq!(12, part2(&example()));
}
